ui/webui/resources/wasm/candle_embedding_gemma/src/lib.rs

```diff
@@ -3,4 +3,986 @@
+            None => Ok(x),
+            Some(bias) => x.broadcast_add(bias),
+        }?;
+
+        if let Some(act) = &self.act {
+            match act {
+                HiddenAct::Gelu => x.gelu(),
+                HiddenAct::Relu => x.relu(),
+                HiddenAct::Silu => x.silu(),
+                HiddenAct::Swiglu => candle_nn::ops::swiglu(&x),
+            }
+        } else {
+            Ok(x)
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Batch {
+    pub input_ids: Vec<u32>,
+    pub token_type_ids: Vec<u32>,
+    pub position_ids: Vec<u32>,
+    pub cumulative_seq_lengths: Vec<u32>,
+    pub max_length: u32,
+    pub pooled_indices: Vec<u32>,
+    pub raw_indices: Vec<u32>,
+}
+
+impl Batch {
+    pub fn len(&self) -> usize {
+        self.cumulative_seq_lengths.len() - 1
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+/// Model type, only embeddings are supported in Gemma.
+#[derive(Debug, PartialEq, Clone)]
+pub enum ModelType {
+    Classifier,
+    Embedding(Pool),
+}
+
+/// Supported pooling strategies.
+#[derive(Debug, PartialEq, Clone, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum Pool {
+    /// Take the CLS token embedding
+    Cls,
+    /// Mean pool across tokens
+    Mean,
+    /// SPLADE (sparse lexical embeddings, not supported in Gemma)
+    Splade,
+    /// Take the last token embedding
+    LastToken,
+}
+
+#[derive(Debug, Clone, PartialEq, Deserialize)]
+pub struct Gemma3Config {
+    pub attention_bias: bool,
+    pub pad_token_id: u32,
+    pub head_dim: Option<usize>,
+    pub hidden_activation: HiddenAct,
+    pub hidden_size: usize,
+    pub intermediate_size: usize,
+    pub max_position_embeddings: usize,
+    pub num_attention_heads: usize,
+    pub num_hidden_layers: usize,
+    pub num_key_value_heads: usize,
+    pub query_pre_attn_scalar: usize,
+    pub rms_norm_eps: f32,
+    pub rope_local_base_freq: f32,
+    pub rope_theta: f32,
+    pub sliding_window: Option<usize>,
+    #[serde(rename(deserialize = "_sliding_window_pattern"))]
+    pub sliding_window_pattern: usize,
+    pub vocab_size: usize,
+}
+
+#[derive(Debug)]
+pub struct Gemma3RMSNorm {
+    weight: Tensor,
+    epsilon: f32,
+}
+
+impl Gemma3RMSNorm {
+    pub fn load(vb: VarBuilder, hidden_size: usize, epsilon: f32) -> candle_core::Result<Self> {
+        Ok(Self {
+            weight: vb.get(hidden_size, "weight").or_else(|_| vb.get(hidden_size, "gamma"))?,
+            epsilon,
+        })
+    }
+
+    pub fn forward(
+        &self,
+        hidden_states: &Tensor,
+        residual: Option<&Tensor>,
+    ) -> candle_core::Result<(Tensor, Tensor)> {
+        let mut hidden_states = hidden_states.clone();
+        let residual_add = if let Some(residual) = residual {
+            let residual_add = hidden_states.add(residual)?;
+            hidden_states = residual_add.clone();
+            residual_add
+        } else {
+            hidden_states.clone()
+        };
```

Comment:
```suggestion
        let residual_add = if let Some(residual) = residual {
            hidden_states.add(residual)?
        } else {
            hidden_states.clone()
        };
        let hidden_states = residual_add.clone();
```

easier to understand and should be equivalent?